### PR TITLE
lsp: Deduplicate result of `queryProperties`

### PR DESCRIPTION
**Pull request by hunger**
When querying properties we walk down the tree of types from the current elements type down through the base types.

We can see the same element several times that way as it may get assigned values on the way down to the definition. So de-duplicate properties in the result and return only the deepest as that is where the property got defined.

### CODE REVIEW

File: tools/lsp/properties.rs [hunk 1]
```diff
@@ -187,7 +187,14 @@ fn get_properties(element: &ElementRc) -> Vec<PropertyInformation> {
         }
     }
 
-    result.sort_by(|a, b| a.name.cmp(&b.name));
+    result.extend(get_reserved_properties()); // Add reserved properties last!
+
+    // We can have duplicate properties that were defined and then changed further up the
+    // Element tree. So we need to remove duplicates.
+    result.sort_by(|a, b| a.name.cmp(&b.name)); // Sort is stable!
+    result.reverse(); // Now the deepest property is first
```
Comment (ogoffart) [inline tools/lsp/properties.rs:195]: I don't understand why you reverse

**Comment by ogoffart** [comment]
Could there be a test?

File: tools/lsp/properties.rs (hunk 1 above)
Comment (hunger) [reply tools/lsp/properties.rs:195]: I go from the actual type of the element down to the base types, adding properties for the current type.

The property is defined by the last entry in that list and gets reassigned values in the earlier versions. I do want the place it is defined at.

Sort is stable, so the definition still happens in the last occurrence of any property.

So I revert the list, moving the definition to the front, where dedup() will leave the definition in place and delete the other occurrences.

Then I reverse back to get the sorting back.

O tried using dedup from itertools, but that does not return a DoupleEndedIterator, so I can not reverse the list based on iterators after dedup.

Having thought about this some more: I think I can avoid one revert: I could sort in reverse order, reverse the Vec and then dedup that.

**Comment by hunger** [comment]
> Could there be a test?

I was expecting this: Working on it:-)

**Comment by hunger** [comment]
New and improved! Now with a simple test!

That made me discover that I was returning file names, not URLs as I should have, so add that.

**Review by tronical — APPROVED** [review]

File: tools/lsp/properties.rs (hunk 1 above)
Comment (ogoffart) [reply tools/lsp/properties.rs]: I see, thanks for the explanation.

File: tools/lsp/properties.rs [hunk 2]
```diff
@@ -65,10 +65,17 @@ fn get_element_properties(element: &Element) -> impl Iterator<Item = PropertyInf
...
         let declared_at = file.as_ref().and_then(|file| {
-            value
-                .type_node()
-                .map(|n| n.text_range().start().into())
-                .map(|p| DeclarationInformation { uri: file.clone(), start_offset: p })
+            value.type_node().map(|n| n.text_range().start().into()).map(|p| {
+                DeclarationInformation {
+                    uri: lsp_types::Url::from_file_path(file)
+                        .unwrap_or_else(|_| {
+                            lsp_types::Url::parse("file:///)")
```
Comment (ogoffart) [inline tools/lsp/properties.rs:72]: There is no default?

**Review by ogoffart — APPROVED** [review]

File: tools/lsp/properties.rs (hunk 2 above)
Comment (hunger) [reply tools/lsp/properties.rs:72]: Nope.

**Pull request merged**